### DISCUSSION TITLE
How to await an event to arise before starting a container.

### DISCUSSION THREAD

**Discussion by Valkozaur** [microsoft/aspire#5176] [Q&A] [answered]
Hello everyone. 
I am currently experimenting with .NET Aspire. So far so good, it seems awesome and I thank Microsoft for this initiative. 

On the project I am working I need to start a container which is depended on other underlying containers to start first.

The container I need to run is Kernel Memory. This container needs an appsettings.json file on startup. It needs to have the connection strings for the Blob Container. I had an issue obtaining them but I got help in this issue: #5115 

Now I fill my appsettings the moment that the connection string is available for my storage emulator using a lifehook.

```
var storage = builder.AddAzureStorage("Storage");

if (builder.Environment.IsDevelopment())
{
    storage.RunAsEmulator();
}

var blobs = storage.AddBlobs("BlobConnection");
var queues = storage.AddQueues("QueueConnection");

blobs.ApplicationBuilder.Services.TryAddLifecycleHook<LifecycleHook>();

await BuildKernelMemorySettings();

builder.AddContainer("kernel-memory", "kernelmemory/service", "latest")
    .WithImageRegistry("docker.io")
    .WithBindMount("kernel.memory.appsettings.Development.json", "/app/appsettings.Development.json", true)
    .WithEnvironment(callback => callback.EnvironmentVariables.Add("ASPNETCORE_ENVIRONMENT", "Development"))
    .WithReference(blobs)
    .WithReference(queues)
    ;
```

```
public class LifecycleHook : IDistributedApplicationLifecycleHook
{
    public async Task AfterEndpointsAllocatedAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken = new CancellationToken())
    {
        var blobStorage = appModel.Resources.OfType<Aspire.Hosting.Azure.AzureBlobStorageResource>().Single();
        var queueStorage = appModel.Resources.OfType<Aspire.Hosting.Azure.AzureQueueStorageResource>().Single();

        var blobConnectionString = await blobStorage.ConnectionStringExpression.GetValueAsync(cancellationToken);
        var queueConnectionString = await queueStorage.ConnectionStringExpression.GetValueAsync(cancellationToken);
        await BuildKernelMemorySettings();
        async Task BuildKernelMemorySettings()
        {
            var kernelMemorySettings = await File.ReadAllTextAsync("kernel.memory.appsettings.Development.json", cancellationToken);
            var jsonNode = JsonNode.Parse(kernelMemorySettings);
            jsonNode["KernelMemory"]["Services"]["AzureBlobs"]["ConnectionString"] = blobConnectionString;
            jsonNode["KernelMemory"]["Services"]["AzureQueues"]["ConnectionString"] = queueConnectionString;

            await File.WriteAllTextAsync("kernel.memory.appsettings.Development.json", jsonNode.ToString(), cancellationToken);
        }
    }
}
```

Though this still does not solve my problem. I need to start the container only when the appsettings is correctly filled, what would be the best way to delay it's start?

**Comment by davidfowl** [comment, marked as answer]
We have this resource eventing model https://learn.microsoft.com/en-us/dotnet/aspire/app-host/eventing#resource-eventing, that lets you block before a resource starts so you can implement custom waiting like this.

You can also use container files (new feature): https://github.com/dotnet/aspire/blob/e1b74bf7f0286c6531e0d58787b1047d8a5aad81/src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs#L204

PS: Here's a big doc on the resource model https://gist.github.com/davidfowl/b408af870d4b5b54a28bf18bffa127e1. It explains some of these primitives.

**Comment by HuntJason** [comment]
I like this. Is there a way to have a dependent resource then WaitFor the EventingResource?

scenario is I have an API service. I want to generate a ClientSecret for my dependent consumer. I can wait for the API service to be ready using the above scenario.

Ideally, I'd like to do the following:
var myService = builder.AddProject<Projects.My_Project>("myProject");

string dependentServiceClientId = "dependentClient";
string dependentServiceClientSecret;
var dependentServiceClientInitialition = builder.Eventing.Subscribe<ResourceReadyEvent>(myService, static (@event, cancellationToken) => {  
    dependentServiceClientSecret = // call MyService to register my dependent service client to generate ClientSecret;
});

var myDependentService = builder.AddProject<Projects.My_Dependent_Project>("dependentProject")
  .WithEnvironment("MyService__ClientId", dependentServiceClientId)
  .WithEnvironment("MyService__ClientSecret", dependentServiceClientSecret)
.WaitFor(dependentServiceClientInitialition);

**Reply by davidfowl** [reply]
There's no eventing resource, you can subscribe to events on a resource. Why do you need to subscribe to anything to generate a client secret? Just generate it using a AddParameter and https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.applicationmodel.generateparameterdefault?view=dotnet-aspire-9.2

**Reply by HuntJason** [reply]
I don't control the initialization of the myService project. Imagine you're pulling a container for a third party software like KeyCloak. There's no way to have it generate a client until after it is started. So, for my dependent service, I need to register it as a client after the service has been started using the keycloak API, THEN pass the generated client secret to myDependentService.

**Reply by davidfowl** [reply]
You can't orchestrate across "runtime state" (state generated from running resources) at the app host level. That requires you to coordinate that state yourself. When you do that, it'll never work outside of run mode (you'll never be able to publish it).
